Cоздание системы управления ЖД путями

qytiz opened the pull request.

qytiz committed:
Update app.rb

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
   end
-  def start_engine
-       puts "Wroom"
+  def add_train(train)
```
Между методами обязательно должна быть пустая строка. Вообще, прочитай [стайл-гайд](https://github.com/arbox/ruby-style-guide/blob/master/README-ruRU.md)

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
   end
-  def beep
-    puts "beep"
+  def show_trains
```
Этот метод не нужен, достаточно просто ридера

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
-  def stop
-    @speed = 0
+  def show_trains_by_type(type)
+    trainType=@trains_list.find_all{ |train| train.type == type}
```
Не используй camelCase для именования переменных - только snake_case

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+  def show_trains_by_type(type)
+    trainType=@trains_list.find_all{ |train| train.type == type}
+    puts "Количество поездов #{type} равно #{trainType.length}"
+    trainType.each{ |train| puts train}
```
Можно просто вывести массив поездов по типу

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
-    @speed = 50
+  def send_train(train)
+    @trains_list.delete(train);
+    
```
Лишняя пустая строка

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+end
+
+class Route
+  attr_reader :first,:optional_s,:last
```
А что значит optional_s? Без разбора кода эта переменная нам ни о чем не говорит

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+  def remove_station(station)
+    @optional_s.delete(station)
+  end
+  def show_stations
```
Лучше просто вернуть массив всех поездов маршрута

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+    @wagon_counter=wagon_counter;
+    @speed=0;
+    @station_number=0;
+    @route=[]
```
В `@route` должен находиться экземпляр класса Route

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
   def current_speed
-    puts @speed
+    print "Текущая скорость =";
+    puts @speed;
```
Ридера на переменную достаточно

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+  end
+  def current_wagon
+    print "Текущее число вагонов =";
+    puts @wagon_counter;
```
Тут тоже достаточно ридера

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+    print "Текущее число вагонов =";
+    puts @wagon_counter;
+  end
+  def wagon(operation)
```
На надо давать аттрибуту возможность управлять логикой метода. Лучше сделать два отдельных метода - на добавление вагона и удаление

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+    end
+  end
+  def route(route)
+  @route<<route.first;
```
А что с отступами?

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+  if @station_number<@route.length
+    @current_station=@route[@station_number];
+  else
+    @station_number-=1;
```
Странно получается. Вначале добавили, потом убавили. Может, лучше сразу проверять что можно ехать?

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+    end
+  end
+
+  def show_route
```
Лучше завести три метода - на возврат текущей, следующей и предыдущей станции. К тому же их можно будет использовать в методах движения

qytiz committed:
Исправления

Теперь станции могут отправлять поезда, а другие станции могут их получать, ваааааааау

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,123 @@
+  end
+ 
+  def route(route)
+    @route_serialized<<route.first
```
Сейчас получается странная вещь - экземпляр маршрута передается в поезд, но не сохраняется, а из него вытаскиваются станции. Зачем? Мы лишаемся преимуществ ооп и размазываем данные по двум классам, да и поезд выполняет ту работу, которую должен делать маршрут. Логичнее сделать у маршрута метод stations, который будет отдавать все станции маршрута, и использовать его из поезда. Поезд должен хранить именно экземпляр маршрута, а не его станции.

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,123 @@
+  end
+
+  def move_forward
+    if get_next_station.is_a?(Station)
```
Если следующая станция - не станция, то это ошибка и программа должна упасть. Подобная проверка бесполезна.

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,123 @@
+
+  def get_ex_station
+  if @station_number>0
+    return @route_serialized[@station_number-1]
```
return тут необязателен

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,123 @@
-  def current_speed
-    puts @speed
+  
+  def show_route
```
Этому методу не место в Train

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,123 @@
+  end
+
+  def get_current_station
+    return @route_serialized[@station_number]
```
return тут необязателен

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,123 @@
+  end
+
+  def add_wagon
+    @wagon_counter+=1
```
Проверь что у тебя везде арифметические операции оформлены по этому [стайлгайду](https://github.com/arbox/ruby-style-guide/blob/master/README-ruRU.md#spaces-operators)

qytiz committed:
Update app.rb

russo-matrosso commented on app.rb:
```diff
@@ -39,81 +43,81 @@ def remove_station(station)
-      @station_number+=1
+      @current_station.remove_train(self)
+      @current_station = get_next_station
+      @station_number += 1
```
А зачем нам индекс станции если мы его можем достать через route?

russo-matrosso commented on app.rb:
```diff
@@ -15,22 +19,22 @@ def show_trains_by_type(type)
 
   def send_train(train)
```
Лучше этот метод убрать. Сейчас получается что у нас за движение отвечает и поезд, и станция. Для станции метода remove_train хватит

russo-matrosso replied:
И все-таки этот метод нужно убрать - он теперь просто обертка. Если хочешь - переименуй remove_train в send_train

russo-matrosso commented on app.rb:
```diff
@@ -39,81 +43,81 @@ def remove_station(station)
     puts @last.name
   end
+  def stations
+    stations = []
```
Присваивание переменной не нужно, она никак потом не используется
И все в одну строку:
```
[@first, *@extra_stations, @last]
```

russo-matrosso commented on app.rb:
```diff
@@ -39,81 +43,81 @@ def remove_station(station)
+    @wagon_counter = wagon_counter
+    @speed = 0
+    @station_number= 0
+    @route
```
Зачем здесь вызов этой переменной:?

russo-matrosso commented on app.rb:
```diff
@@ -39,81 +43,81 @@ def remove_station(station)
-    end
   end
 
   def get_ex_station
```
get_previous_station
Ну зачем станцию "бывшей" называть...

qytiz committed:
Update app.rb

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+  end
+  end
+
+  def get_current_station
```
Я бы избавился от слова get_ в названиях метода. И конкретно на текущую станцию можно просто повесить attr_reader

russo-matrosso commented on app.rb:
```diff
@@ -1,20 +1,117 @@
+
+  def move_backward
+      @current_station.remove_train(self)
+      @current_station = get_ex_station
```
А где метод get_ex_station? может, get_previous_station?

qytiz committed:
Update app.rb
